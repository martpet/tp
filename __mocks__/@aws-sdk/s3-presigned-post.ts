export const createPresignedPost = vi.fn().mockResolvedValue('dummyPresignedPost');
