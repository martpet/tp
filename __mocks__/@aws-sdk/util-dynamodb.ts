export const unmarshall = vi
  .fn()
  .mockName('unmarshall')
  .mockReturnValue('dummyUnmarshalledItems');
