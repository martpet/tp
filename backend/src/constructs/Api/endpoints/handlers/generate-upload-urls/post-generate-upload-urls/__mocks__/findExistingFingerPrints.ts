export const findExistingFingerPrints = vi
  .fn()
  .mockName('findExistingFingerPrints')
  .mockResolvedValue(['dummyFingerprint1']);
