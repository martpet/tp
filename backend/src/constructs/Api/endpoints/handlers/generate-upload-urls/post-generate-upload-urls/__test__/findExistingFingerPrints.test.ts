import { BatchGetCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { itResolves, itSendsAwsCommand, photosTableOptions } from 'lambda-layer';

import { findExistingFingerPrints } from '../findExistingFingerPrints';

vi.mock('@aws-sdk/util-dynamodb');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [['fingerprint1']] as Parameters<typeof findExistingFingerPrints>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(BatchGetCommand).resolves({
    Responses: {
      [photosTableOptions.tableName]: [
        { fingerprint: 'existingFingerprint1' },
        { fingerprint: 'existingFingerprint1' },
      ],
    },
  });
});

describe('findExistingFingerPrints', () => {
  itSendsAwsCommand(BatchGetCommand, ddbMock, findExistingFingerPrints, args);
  itResolves(findExistingFingerPrints, args);

  describe('when called with 101 items', async () => {
    const argsClone = structuredClone(args);
    argsClone[0] = Array(101);

    it('sends "BatchGetCommand" 2 times', async () => {
      await findExistingFingerPrints(...argsClone);
      expect(ddbMock.commandCalls(BatchGetCommand).length).toBe(2);
    });
  });

  describe('when "BatchGetCommand" response is missing "Responses"', () => {
    beforeEach(() => {
      ddbMock.on(BatchGetCommand).resolves({});
    });
    itResolves(findExistingFingerPrints, args);
  });

  describe('when "BatchGetCommand" response has "UnprocessedKeys"', () => {
    beforeEach(() => {
      ddbMock
        .on(BatchGetCommand)
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ fingerprint: 'existingFingerprint3' }],
          },
          UnprocessedKeys: {
            [photosTableOptions.tableName]: { Keys: [{ dummyUnprocessedKey: '' }] },
          },
        })
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ fingerprint: 'existingFingerprint4' }],
          },
        });
    });

    it('sends "BatchGetCommand" a second time with correct args', async () => {
      await findExistingFingerPrints(...args);
      expect(ddbMock.commandCalls(BatchGetCommand)[1].args[0].input).toMatchSnapshot();
    });

    itResolves(findExistingFingerPrints, args);
  });
});
