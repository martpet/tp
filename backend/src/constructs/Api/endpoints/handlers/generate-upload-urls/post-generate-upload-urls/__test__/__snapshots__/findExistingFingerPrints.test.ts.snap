// Vitest Snapshot v1

exports[`findExistingFingerPrints > resolves with a correct value 1`] = `
[
  "existingFingerprint1",
  "existingFingerprint1",
]
`;

exports[`findExistingFingerPrints > sends "BatchGetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        {
          "fingerprint": "fingerprint1",
        },
      ],
    },
  },
}
`;

exports[`findExistingFingerPrints > when "BatchGetCommand" response has "UnprocessedKeys" > resolves with a correct value 1`] = `
[
  "existingFingerprint3",
  "existingFingerprint4",
]
`;

exports[`findExistingFingerPrints > when "BatchGetCommand" response has "UnprocessedKeys" > sends "BatchGetCommand" a second time with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        "dummyUnmarshalledItems",
      ],
    },
  },
}
`;

exports[`findExistingFingerPrints > when "BatchGetCommand" response is missing "Responses" > resolves with a correct value 1`] = `[]`;
